app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_iaa.rb

```diff
@@ -0,0 +1,194 @@
+        }
+
+        sql = format(<<~SQL, params)
+          WITH subquery AS (%{subquery})
+          SELECT
+            billing_month_logs.year_month
+          , billing_month_logs.ial
+          , %{iaa} AS iaa
+          , %{iaa_start_date} AS iaa_start_date
+          , %{iaa_end_date} AS iaa_end_date
+          , %{select_clause}
+          FROM
+            subquery billing_month_logs
+          WHERE
+            %{where_clause}
+          GROUP BY
+            billing_month_logs.year_month
+          , billing_month_logs.ial
+        SQL
+
+        ActiveRecord::Base.connection.execute(sql)
+      end
+
+      # @return [String]
+      def full_month_subquery(issuers:, full_months:)
+        params = {
+          issuers: issuers,
+          year_months: full_months.map { |r| r.begin.strftime('%Y%m') },
+        }.transform_values { |value| quote(value) }
+
+        full_month_subquery = format(<<~SQL, params)
+          SELECT
+            monthly_sp_auth_counts.user_id
+          , monthly_sp_auth_counts.year_month
+          , monthly_sp_auth_counts.auth_count
+          , monthly_sp_auth_counts.ial
+          FROM
+            monthly_sp_auth_counts
+          WHERE
+                monthly_sp_auth_counts.issuer IN %{issuers}
+            AND monthly_sp_auth_counts.year_month IN %{year_months}
+        SQL
+      end
+
+      # @return [Array<String>]
+      def partial_month_subqueries(issuers:, partial_months:)
+        partial_months.map do |month_range|
+          params = {
+            range_start: month_range.begin,
+            range_end: month_range.end,
+            year_month: month_range.begin.strftime('%Y%m'),
+            issuers: issuers,
+          }.transform_values { |value| quote(value) }
+
+
+          format(<<~SQL, params)
+            SELECT
+              sp_return_logs.user_id
+            , %{year_month} AS year_month
+            , COUNT(sp_return_logs.id) AS auth_count
+            , sp_return_logs.ial
+            FROM sp_return_logs
+            INNER JOIN service_providers
+              ON sp_return_logs.issuer = service_providers.issuer
+            WHERE
+                  sp_return_logs.requested_at BETWEEN %{range_start} AND %{range_end}
+              AND sp_return_logs.returned_at IS NOT NULL
+              AND sp_return_logs.requested_at BETWEEN
+                    service_providers.iaa_start_date AND service_providers.iaa_end_date
+              AND service_providers.issuer IN %{issuers}
+            GROUP BY
+              sp_return_logs.user_id
+            , sp_return_logs.ial
+          SQL
+        end
+      end
+
+      def quote(value)
+        if value.is_a?(Array)
+          "(#{value.map { |v| ActiveRecord::Base.connection.quote(v) }.join(', ')})"
+        else
+          ActiveRecord::Base.connection.quote(value)
+        end
+      end
+
+      # @param [Range<Date>] date_range
+      # @return [Array<Range<Date>>]
+      def months(date_range)
```

Comment:
could documentation be added to describe what's being built in this?

Reply:
Added in 1bc14550d92d8681761f55a3cc5ebda174553b13, PTAL

Reply:
yep, I get it now